apache/trafficcontrol
# TP: improved snapshot diff performance/experience

## What does this PR (Pull Request) do?

- [x] This PR fixes #3275 

Currently, when a cdn snapshot gets too large, the current mechanism used to show the differences between the current snapshot and the pending snapshot struggles and is not performant. This improvement leverages a must faster library (https://www.npmjs.com/package/fast-json-patch) for identifying differences between 2 large json objects. 

However, because of the use of a different library with a different output, the UI experience changed somewhat. Differences (add, remove or update) are now captured in a table that looks as such:

![image](https://user-images.githubusercontent.com/251272/57107780-2b8a2200-6cee-11e9-86ba-03642e452f32.png)

![image](https://user-images.githubusercontent.com/251272/57107948-9b001180-6cee-11e9-92fb-f95b93305b2c.png)

## Which Traffic Control components are affected by this PR?

- Traffic Portal

## What is the best way to verify this PR?

- Navigate to https://tp.domain.com/#!/cdns/:cdn_id/config/changes
- View any outstanding differences between the current and pending snapshot. If there are none, you can:

a. modify the attributes of a delivery service (note: not all ds attributes are captured in a snapshot)
b. add/removes caches from a delivery service
c. modify the values of some CRConfig.json parameters (go to https://tp.domain.com/#!/parameters and search for CRConfig.json)
d. change the status of caches. for example, from reported to admin_down
e. change the status of some traffic routers. for example, from online to admin_down
f. change the attributes of some edge cache groups where type=EDGE_LOC
g. change the attributes of some TR cache groups where type=CCR

- Now go back to https://tp.domain.com/#!/cdns/:cdn_id/config/changes and review the differences.
- Optional: Perform snapshot. this will set current snapshot == pending snapshot, therefore, your differences will go away and you won't see any changes at https://tp.domain.com/#!/cdns/:cdn_id/config/changes

Note: although maybe perceived as beneficial, UI tests would probably not be worth the cost.
Note: there was never any documentation regarding the snapshot diff functionality of TP so none to edit. 

## The following criteria are ALL met by this PR
